src/fluree/db/json_ld/shacl.cljc

```diff
@@ -183,15 +183,65 @@
     (coalesce-validation-results results)))
 
 (defn validate-value-properties
-  ;; TODO: Only supports 'in' so far. Add the others.
-  [{:keys [in logical-constraint] :as _p-shape} p-flakes]
-  (let [results (for [flake p-flakes
-                      :let [[val] (flake-value flake)
-                            in-set (set in)]]
-                  (if (in-set val)
-                    [true (str "sh:not sh:in: value " val " must not be one of " in)]
-                    [false (str "sh:in: value " val " must be one of " in)]))]
-    (coalesce-validation-results results logical-constraint)))
+  [{:keys [in has-value datatype nodekind logical-constraint] :as _p-shape} p-flakes]
+  (let [in-results (when in
+                     (if (every? #(contains? (set in) (flake/o %)) p-flakes)
+                       [true (str "sh:not sh:in: value " val " must not be one of " in)]
+                       [false (str "sh:in: value " val " must be one of " in)]))
+        has-value-results (when has-value
+                            (if (some #(= (flake/o %) has-value) p-flakes)
+                              [true (str "sh:not sh:hasValue: at least one value must not be " has-value)]
```

Comment:
I tend to get kind of mixed up with the `sh:not` cases, but I'm not sure this is the right message here. I thought it would be be that none of the values can have the `has-value`?